src/printf.c

```diff
@@ -0,0 +1,101 @@
+#include "printf.h"
+#include "uart.h"
+#include "lcd.h"
+
+#include <stdarg.h>
+#include <stdint.h>
+
+static void putchar(writer *stream, char c)
+{
+    stream->write(stream, &c, 1);
+}
+
+static void putstr(writer *stream, char *str)
+{
+    for (int i = 0; str[i]; i++)
+        putchar(stream, str[i]);
```

Comment:
This looks rather inefficient. Why not use `stream->write(stream, str, strlen(str));`?